server/src/main/java/org/elasticsearch/search/rank/context/RankFeaturePhaseRankCoordinatorContext.java

```diff
@@ -0,0 +1,93 @@
+        this.rankWindowSize = rankWindowSize;
+    }
+
+    /**
+     * Computes the updated scores for a list of features (i.e. document-based data). We also pass along an ActionListener
+     * that should be called with the new scores, and will continue execution to the next phase
+     */
+    protected abstract void computeScores(RankFeatureDoc[] featureDocs, ActionListener<float[]> scoreListener);
+
+    /**
+     * This method is responsible for ranking the global results based on the provided rank feature results from each shard.
+     * <p>
+     * We first start by extracting ordered feature data through a {@code List<RankFeatureDoc>}
+     * from the provided rankSearchResults, and then compute the updated score for each of the documents.
+     * Once all the scores have been computed, we sort the results, perform any pagination needed, and then call the `onFinish` consumer
+     * with the final array of {@link ScoreDoc} results.
+     *
+     * @param rankSearchResults a list of rank feature results from each shard
+     * @param rankListener      a rankListener to handle the global ranking result
+     */
+    public void rankGlobalResults(List<RankFeatureResult> rankSearchResults, ActionListener<RankFeatureDoc[]> rankListener) {
+        // extract feature data from each shard rank-feature phase result
+        RankFeatureDoc[] featureDocs = extractFeatureDocs(rankSearchResults);
+
+        // generate the final `topResults` paginated results, and pass them to fetch phase through the `rankListener`
+        computeScores(featureDocs, rankListener.delegateFailureAndWrap((listener, scores) -> {
+            for (int i = 0; i < featureDocs.length; i++) {
+                featureDocs[i].score = scores[i];
+            }
+            listener.onResponse(featureDocs);
+        }));
+    }
+
+    /**
+     * Ranks the provided {@link RankFeatureDoc} array and paginates the results based on the `from` and `size` parameters.
+     */
+    public RankFeatureDoc[] rankAndPaginate(RankFeatureDoc[] rankFeatureDocs) {
+        Arrays.sort(rankFeatureDocs, Comparator.comparing((RankFeatureDoc doc) -> doc.score).reversed());
+        RankFeatureDoc[] topResults = new RankFeatureDoc[Math.max(0, Math.min(size, rankFeatureDocs.length - from))];
+        for (int rank = 0; rank < topResults.length; ++rank) {
+            topResults[rank] = rankFeatureDocs[from + rank];
+            topResults[rank].rank = from + rank + 1;
+        }
+        return topResults;
+    }
+
+    private RankFeatureDoc[] extractFeatureDocs(List<RankFeatureResult> rankSearchResults) {
+        List<RankFeatureDoc> docFeatures = new ArrayList<>();
+        for (RankFeatureResult rankFeatureResult : rankSearchResults) {
+            RankFeatureShardResult shardResult = rankFeatureResult.shardResult();
+            docFeatures.addAll(Arrays.stream(shardResult.rankFeatureDocs).filter(x -> Objects.nonNull(x.featureData)).toList());
```

Comment:
Maybe do this a little more efficiently and save the allocation of the intermediary list just to `addAll` it afterwards?
I'd also generally try to stay away from using streams in nested loops, performance wise the allocations start to hurt and you could just as well build this thing with just normal looping over the array which won't allocate a thing?

Reply:
Yeap, fair point. Updated to simply iterate over the array.